Custom memory

pedro-m-dietrich opened the pull request:
Create chatbot with custom memory.

The chatbot uses the last 5 messages of the conversation and a context that stores a resume of the entire conversation and information about the user.

This avoids excessive token usage for longer conversations, while keeping the relevant information available.

pedro-m-dietrich committed:
Move `create_handoff_tool()` to `chat_config.py`

The `create_handoff_tool()` function can be used in multiple chatbots,
therefore it was moved to the `chat_config.py` file to avoid code repetition.

pedro-m-dietrich committed:
Save context in chat history and DB

The `ChatHistory` class can now save the conversation context as a
string.

The chat context can be stored in the database as a text field.

pedro-m-dietrich committed:
Create a chatbot with custom memory

A chatbot with a custom memory has been created.

The chatbot only uses the last 5 messages of the chat and a context message
that tries to resume the rest of the conversation and save user data.

NixLoar commented on scripts/08_langgraph_custom_memory.py:
```diff
@@ -0,0 +1,220 @@
+        Router's structured response to decide which multi-agent system will be used to answer the user.
+        The `decision` must be either `RESEARCH` or `MATH`. Use `RESEARCH` by default.
+    """
+    decision: ChatbotSystems
```
sugestão: você pode definir um valor default assim em vez de deixar pra LLM resolver

```suggestion
    decision: ChatbotSystems = Field(default=ChatbotSystems.RESEARCH)
```

NixLoar commented on scripts/08_langgraph_custom_memory.py:
```diff
@@ -0,0 +1,220 @@
+        Keep the context data as short as possible, while keeping the important information.
+    """
+    chat_summary: str = Field(..., description='Summary of the current conversation between the user and the chatbot system.')
+    user_data: str = Field(..., description='Relevant information about the user.')
```
a ideia é essa mesmo, mas vc pode (e geralmente queremos ser) mais explicito com os dados que queremos salvar. Por exemplo, criando uma tipagem pro user_data:

```
class UserData(BaseModel):
    name: Optional[str] = Field(None, description="User's name if mentioned.")
    age: Optional[int] = Field(None, description="User's age if specified.")
    gender: Optional[str] = Field(None, description="User's gender if stated.")
```
    
E no user_data alterar pra esse tipo:
```suggestion
    user_data: UserData = Field(..., description='Relevant information about the user.')
```

pedro-m-dietrich committed:
Separate multi-agent systems in subgraphs

Each multi-agent system now has its own subgraph, to improve the graph
flux and isolate the systems properly.

Japoncio3k reviewed (approved):
Fazer dessa forma resolveu o problema que tava tendo?

pedro-m-dietrich commented:
> Fazer dessa forma resolveu o problema que tava tendo?

Sim, pelo que eu testei, o `context_agent` não está sendo repetido.
